<a href="https://colab.research.google.com/github/elvano22/News-Classification/blob/main/Automatic_News_Classification.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Crawling dan Scraping

In [ ]:
from bs4 import BeautifulSoup
import requests
import pandas as pd
import datetime

## Crawling kompas.com

In [ ]:
links = []

indexes = ['travel','bola','nasional', 'umkm']

for i in range (1,300):
  if len(links) > 100: break;

  for index in indexes:
    if len(links) > 100: break;

  # URL dari Kompas.com
  url = f'https://indeks.kompas.com/?site={index}&page={i}'

  # Mengambil halaman dengan requests
  response = requests.get(url)

  # Parsing konten halaman menggunakan BeautifulSoup
  soup = BeautifulSoup(response.text, 'html.parser')

  # Mencari semua elemen berita utama, misalkan menggunakan class yang spesifik pada elemen berita
  articles = soup.find_all('div', class_='articleItem')

  # Mencari link dari setiap berita
  for article in articles:
    link = article.find('a', class_='article-link')['href']
    links.append(link)

# Melihat 5 link pertama
links[:5]

['https://umkm.kompas.com/read/2024/04/24/210000483/berdayakan-umkm-orang-asli-papua-pemkab-sorong-anggarkan-rp1-7-miliar',
 'https://umkm.kompas.com/read/2024/04/24/185248683/dari-driver-ojek-online-olan-sukses-jadi-juragan-loyang-dengan-bantuan-kur',
 'https://umkm.kompas.com/read/2024/04/24/155544583/wujudkan-passion-di-bidang-fesyen-frida-aulia-bangun-bisnis-hingga-kenalkan',
 'https://umkm.kompas.com/read/2024/04/24/150500383/3-cara-mengatasi-ulasan-negatif-di-platform-online-pelaku-usaha-harus-tahu',
 'https://umkm.kompas.com/read/2024/04/24/122148783/cara-soes-surgawi-manfaatkan-kuis-berhadiah-untuk-strategi-promosi']

In [ ]:
len(links)

105

## Scraping kompas.com

In [ ]:
labels =[]
bodies = []
medias = []
for link in links:
  # Mengambil halaman dengan requests
  response = requests.get(f'{link}?page=all#page')
  if response.status_code != 200: continue

  # Parsing konten halaman menggunakan BeautifulSoup
  soup = BeautifulSoup(response.text, 'html.parser')

  # Mencari semua isi berita
  body_temp = soup.find_all('p')
  body_temp
  body_texts=[]
  for x in body_temp:
    if not x.has_attr('class'):
      body_texts.append(x.get_text().replace("\xa0", " "))

  # Mencari label
  labels_temp = soup.find_all('li', class_='breadcrumb__item')

  if len(labels_temp) == 0 or len(body_texts)<2: continue
  labels.append(labels_temp[len(labels_temp)-1].get_text())
  bodies.append(" ".join(body_texts))
  medias.append("Kompas")

In [ ]:
# Membuat dataframe
df = pd.DataFrame({
    'Label': labels,
    'Media': medias,
    'Body': bodies
})
df

Label   Media                                               Body
0         Program  Kompas  Editor  SORONG, KOMPAS.com - Pemerintah Kabupa...
1    Jagoan Lokal  Kompas  Penulis JAKARTA, KOMPAS.com – Suara lakban yan...
2    Jagoan Lokal  Kompas  Tim Redaksi  BOGOR, KOMPAS.com - Sebuah galeri...
3        Training  Kompas  Tim Redaksi  KOMPAS.com - Jika Anda menggunaka...
4        Training  Kompas  Tim Redaksi TANGERANG, KOMPAS.com - Siang hari...
..            ...     ...                                                ...
100      Training  Kompas  Tim Redaksi  KOMPAS.com - Industri fesyen menj...
101  Jagoan Lokal  Kompas  Tim Redaksi  BOGOR, KOMPAS.com - Bisnis oleh-o...
102  Jagoan Lokal  Kompas  Tim Redaksi JAKARTA, KOMPAS.com - Saat ini, su...
103      Training  Kompas  Tim Redaksi KOMPAS.com - Negosiasi bisnis meru...
104      Training  Kompas  Tim Redaksi  JAKARTA, KOMPAS.com – Di tengah p...

[105 rows x 3 columns]

## Crawling liputan6.com

In [ ]:
links_2=[]

# Tanggal awal yang ingin Anda mulai dari tanggal terbaru
start_date = datetime.date.today()

indexes = ['showbiz', 'bola', 'bisnis', 'pemilu']

for i in range (1,300):
  if len(links_2) > 100: break;

  for index in indexes:
    if len(links_2) > 100: break;

    # URL dari liputan6.com
    current_date = start_date - datetime.timedelta(days=i)
    url = f'https://www.liputan6.com/{index}/indeks/{current_date.year}/{current_date.month:02d}/{current_date.day:02d}'

    # Mengambil halaman dengan requests
    response = requests.get(url)

    # Parsing konten halaman menggunakan BeautifulSoup
    soup = BeautifulSoup(response.text, 'html.parser')

    # Mencari semua elemen berita utama, misalkan menggunakan class yang spesifik pada elemen berita
    articles = soup.find_all('article', class_='articles--rows--item')

    # Mencari link dari setiap berita
    for article in articles:
      link = article.find('a', class_='articles--rows--item__title-link')['href']
      links_2.append(link)

# Melihat 5 link pertama
links_2[:5]

['https://www.liputan6.com/showbiz/read/5579978/incubus-nostalgia-dengan-penggemarnya-di-jakarta-saat-manggung-untuk-kali-keempat',
 'https://www.liputan6.com/showbiz/read/5579814/jangan-lewatkan-episode-tertawan-hati-yang-akan-mengubah-semua-cerita-saksikan-hanya-di-sctv',
 'https://www.liputan6.com/showbiz/read/5579950/selebgram-chandrika-chika-ditangkap-polisi-terkait-kasus-narkoba-bersama-pemuda-pemudi-lainnya',
 'https://www.liputan6.com/showbiz/read/5579583/episode-penuh-ketegangan-bidadari-surgamu-sakinah-dalam-bahaya-saksikan-hanya-di-sctv',
 'https://www.liputan6.com/showbiz/read/5579507/trailer-baru-deadpool-amp-wolverine-meluncur-ini-5-hal-yang-bakalan-bikin-seru-termasuk-mayat-raksasa-ant-man']

In [ ]:
len(links_2)

109

## Scraping liputan6.com

In [ ]:
labels =[]
bodies = []
medias = []
for link in links_2:
  # Mengambil halaman dengan requests
  response = requests.get(link)
  if response.status_code != 200: continue

  # Parsing konten halaman menggunakan BeautifulSoup
  soup = BeautifulSoup(response.text, 'html.parser')

  # Mencari semua isi berita
  body_temp = soup.find_all('p')
  body_temp
  body_texts=[]
  for x in body_temp:
    if not x.has_attr('class'):
      body_texts.append(x.get_text().replace("\xa0", " "))

  # Mencari label
  labels_temp = soup.find_all('li', class_='read-page--breadcrumb--item')

  if len(labels_temp) == 0 or len(body_texts)<2: continue
  labels.append(labels_temp[len(labels_temp)-1].get_text())
  bodies.append(" ".join(body_texts))
  medias.append("Liputan6")

In [ ]:
# Membuat dataframe
df_liputan6 = pd.DataFrame({
    'Label': labels,
    'Media': medias,
    'Body': bodies
})
df_liputan6

Label     Media  \
0                 Musik  Liputan6   
1               ShowBiz  Liputan6   
2            Selebritis  Liputan6   
3               ShowBiz  Liputan6   
4           Blink-Blink  Liputan6   
..                  ...       ...   
100      Liga Indonesia  Liputan6   
101                Bola  Liputan6   
102  Liga Internasional  Liputan6   
103              Sports  Liputan6   
104        Liga Inggris  Liputan6   

                                                  Body  
0    Incubus menggelar konser bertajuk Asia Tour 20...  
1    Sinetron Tertawan Hati SCTV tampilkan episode ...  
2    Penangkapan enam pemuda pemudi termasuk selebg...  
3    Sinetron SCTV Bidadari Surgamu semakin seru da...  
4    Berikut hal-hal yang bakal menjadi daya tarik ...  
..                                                 ...  
100  Timnas Indonesia U-23 harus merelakan kehilang...  
101  Penjaga Gawang dan Manajer Barcelona ungkap ke...  
102  Mayoritas dari 830 gol Messi terjadi saat masi...  
103  Delapan pertandingan game pertama play-off NBA...  
104  Nacho telah bersama Real Madrid sejak usia 11 ...  

[105 rows x 3 columns]

## Crawling detik.com

In [ ]:
links_3=[]

indexes = ['https://sport.detik.com/', 'https://travel.detik.com/', 'https://finance.detik.com/ekonomi-bisnis', 'https://news.detik.com/pemilu']

for i in range (1,300):
  if len(links_3) > 100: break;

  for index in indexes:
    if len(links_3) > 100: break;

  # URL dari detik.com
  url = f'{index}/indeks/{i}'

  # Mengambil halaman dengan requests
  response = requests.get(url, allow_redirects=False)

  # Parsing konten halaman menggunakan BeautifulSoup
  soup = BeautifulSoup(response.text, 'html.parser')

  # Mencari semua elemen berita utama, misalkan menggunakan class yang spesifik pada elemen berita
  articles = soup.find_all('article', class_='list-content__item')

  # Mencari link dari setiap berita
  for article in articles:
    link = article.find('a', class_='media__link')['href']
    links_3.append(link)

# Melihat 5 link pertama
links_3[:5]

['https://news.detik.com/pemilu/d-7309304/ketum-hipakad-puji-kenegarawanan-prabowo-yang-resmi-jadi-presiden-terpilih',
 'https://news.detik.com/pemilu/d-7309294/diusung-pks-jadi-bakal-cawalkot-depok-imam-bicara-amanah-berat',
 'https://news.detik.com/pemilu/d-7309272/pks-resmi-usung-imam-budi-hartono-maju-pemilihan-wali-kota-depok',
 'https://news.detik.com/pemilu/d-7309238/soroti-dissenting-opinion-3-hakim-mk-hnw-harap-pemilu-bisa-lebih-baik',
 'https://news.detik.com/pemilu/d-7309232/prabowo-tinggalkan-istana-usai-lebih-dari-2-jam-bertemu-jokowi']

In [ ]:
len(links_3)

120

## Scraping detik.com

In [ ]:
labels =[]
bodies = []
medias = []
for link in links_3:
  # Mengambil halaman dengan requests
  response = requests.get(link)
  if response.status_code != 200: continue

  # Parsing konten halaman menggunakan BeautifulSoup
  soup = BeautifulSoup(response.text, 'html.parser')

  # Mencari semua isi berita
  body_temp = soup.find_all('p')
  body_temp
  body_texts=[]
  for x in body_temp:
    if not x.has_attr('class'):
      body_texts.append(x.get_text().replace("\xa0", " "))

  # Mencari label
  labels_temp = soup.select('div.page__breadcrumb a')

  # Menambah media
  if len(labels_temp) == 0 or len(body_texts)<2: continue
  labels.append(labels_temp[len(labels_temp)-1].get_text().strip())
  bodies.append(" ".join(body_texts))
  medias.append("Detik")

In [ ]:
# Membuat dataframe
df_detik = pd.DataFrame({
    'Label': labels,
    'Media': medias,
    'Body': bodies
})
df_detik

Label  Media                                               Body
0    Pemilu  Detik  Ketum Himpunan Putra-Putri Keluarga Angkatan D...
1    Pemilu  Detik  Imam Budi Hartono resmi diusung PKS untuk maju...
2    Pemilu  Detik  PKS menyerahkan SK untuk bakal calon wali kota...
3    Pemilu  Detik  Wakil Ketua MPR RI Hidayat Nur Wahid (HNW) mem...
4    Pemilu  Detik  Presiden terpilih Prabowo Subianto telah menin...
..      ...    ...                                                ...
115  Pemilu  Detik  KPK menilai jika mengimbau pelarangan pemberia...
116  Pemilu  Detik  Mahkamah Konstitusi (MK) menolak gugatan Anies...
117  Pemilu  Detik  Wali Kota Medan Bobby Nasution mengungkapkan k...
118  Pemilu  Detik  Ketua Harian Gerindra Dasco tak banyak membebe...
119  Pemilu  Detik  Ketua Umum PKB, Muhaimin Iskandar atau Cak Imi...

[120 rows x 3 columns]

## Concat all dataframe from the scraping process



In [ ]:
df_merged = pd.concat([df, df_liputan6, df_detik], ignore_index=True)
df_merged

Label   Media                                               Body
0         Program  Kompas  Editor  SORONG, KOMPAS.com - Pemerintah Kabupa...
1    Jagoan Lokal  Kompas  Penulis JAKARTA, KOMPAS.com – Suara lakban yan...
2    Jagoan Lokal  Kompas  Tim Redaksi  BOGOR, KOMPAS.com - Sebuah galeri...
3        Training  Kompas  Tim Redaksi  KOMPAS.com - Jika Anda menggunaka...
4        Training  Kompas  Tim Redaksi TANGERANG, KOMPAS.com - Siang hari...
..            ...     ...                                                ...
325        Pemilu   Detik  KPK menilai jika mengimbau pelarangan pemberia...
326        Pemilu   Detik  Mahkamah Konstitusi (MK) menolak gugatan Anies...
327        Pemilu   Detik  Wali Kota Medan Bobby Nasution mengungkapkan k...
328        Pemilu   Detik  Ketua Harian Gerindra Dasco tak banyak membebe...
329        Pemilu   Detik  Ketua Umum PKB, Muhaimin Iskandar atau Cak Imi...

[330 rows x 3 columns]

## Standardizing labels in the data

In [ ]:
df_merged['Label'].unique()

array(['Program', 'Jagoan Lokal', 'Training', 'Musik', 'ShowBiz',
       'Selebritis', 'Blink-Blink', 'K-Pop', 'Liga Inggris', 'Sports',
       'Liga Indonesia', 'Bola', 'Liga Italia', 'Ekonomi', 'Bank',
       'Bisnis', 'Pemilu', 'Liga Internasional'], dtype=object)

In [ ]:
# Definisi daftar untuk masing-masing kategori
politik = ['politik', 'nasional', 'pemilu', 'pemerintah', 'pemilu']
olahraga = ['timnas indonesia', 'liga italia', 'bulutangkis', 'olahraga', 'liga inggris', 'liga indonesia', 'badminton', 'sports', 'sport', 'bola', 'liga italia', 'liga internasional']
bisnis = ['e-business', 'swasta', 'bisnis', 'spend smart', 'earn smart', 'smartpreneur', 'housing', 'ekonomi', 'program', 'jagoan lokal', 'training', 'ekonomi', 'bank']
hiburan = ['hype', 'whats new', 'stori', 'feel good', 'hiburan', 'travel', 'travel update', 'aksesoris', 'jalan jalan', 'travel update', 'tempat makan', 'resep', 'game', 'tips kuliner', 'musik', 'showbiz', 'selebritis', 'blink-blink', 'k-pop']

# Fungsi untuk menentukan kategori berdasarkan label
def determine_category(label):
  if label is None:
    return None
  label = label.lower()
  if label in politik:
    return 'politik'
  elif label in olahraga:
    return 'olahraga'
  elif label in bisnis:
    return 'bisnis'
  elif label in hiburan:
    return 'hiburan'
  else:
    return None

# Terapkan fungsi pada kolom label
df_merged['Label'] = df_merged['Label'].apply(determine_category)

# Membuang baris yang kategorinya menjadi None setelah penggantian
df_merged.dropna(subset=['Label'], inplace=True)
df_merged.reset_index(drop=True, inplace=True)

# Tampilkan hasil DataFrame yang telah diperbarui
df_merged

Label   Media                                               Body
0     bisnis  Kompas  Editor  SORONG, KOMPAS.com - Pemerintah Kabupa...
1     bisnis  Kompas  Penulis JAKARTA, KOMPAS.com – Suara lakban yan...
2     bisnis  Kompas  Tim Redaksi  BOGOR, KOMPAS.com - Sebuah galeri...
3     bisnis  Kompas  Tim Redaksi  KOMPAS.com - Jika Anda menggunaka...
4     bisnis  Kompas  Tim Redaksi TANGERANG, KOMPAS.com - Siang hari...
..       ...     ...                                                ...
325  politik   Detik  KPK menilai jika mengimbau pelarangan pemberia...
326  politik   Detik  Mahkamah Konstitusi (MK) menolak gugatan Anies...
327  politik   Detik  Wali Kota Medan Bobby Nasution mengungkapkan k...
328  politik   Detik  Ketua Harian Gerindra Dasco tak banyak membebe...
329  politik   Detik  Ketua Umum PKB, Muhaimin Iskandar atau Cak Imi...

[330 rows x 3 columns]

In [ ]:
df.to_csv("Rawdata_UTS.csv", index=False)

In [ ]:
df_merged['Label'].value_counts()

Label
politik     128
bisnis      125
olahraga     39
hiburan      38
Name: count, dtype: int64

# Text Preprocessing

In [ ]:
!pip install Sastrawi

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 209.7/209.7 kB 5.5 MB/s eta 0:00:00


In [ ]:
import re
import nltk
from Sastrawi.Stemmer.StemmerFactory import StemmerFactory
from nltk.tokenize import word_tokenize
from Sastrawi.StopWordRemover.StopWordRemoverFactory import StopWordRemoverFactory
from Sastrawi.Stemmer.StemmerFactory import StemmerFactory

In [ ]:
nltk.download('punkt')
nltk.download('stopwords')
nltk.download('wordnet')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package wordnet to /root/nltk_data...


True

## Cleansing

In [ ]:
def clean_text(text):
  # Menghapus kata "Advertisement" (case insensitive)
  text = re.sub(r'\bAdvertisement\b', '', text)
  # Menghapus kalimat terakhir dari "* Fakta atau Hoaks?" sampai akhir
  text = re.sub(r'\* Fakta atau Hoaks\?[\s\S]*$', '', text)
  # Mengubah semua teks menjadi lowercase
  text = text.lower()
  # Membuat pattern regex untuk menemukan "KOMPAS.com - " atau "KOMPAS.com — " tanpa memperdulikan case
  pattern = re.compile(r'kompas\.com\s*[-—]\s*')
  # Menghapus teks sebelum pattern
  cleaned = pattern.split(text, 1)[1] if pattern.search(text) else text
  # Menghapus bagian akhir mulai dari " Copyright"
  cleaned = re.split(r'\s*copyright', cleaned, 1)[0]
  # Menghilangkan semua tanda baca
  cleaned = re.sub(r'[^\w\s]', '', cleaned)
  # Menghapus semua angka
  cleaned = re.sub(r'\d+', '', cleaned)
  return cleaned

# Menerapkan fungsi ke kolom teks
df_merged['Cleaned Text'] = df_merged['Body'].apply(clean_text)
df_merged

Label   Media                                               Body  \
0     bisnis  Kompas  Editor  SORONG, KOMPAS.com - Pemerintah Kabupa...   
1     bisnis  Kompas  Penulis JAKARTA, KOMPAS.com – Suara lakban yan...   
2     bisnis  Kompas  Tim Redaksi  BOGOR, KOMPAS.com - Sebuah galeri...   
3     bisnis  Kompas  Tim Redaksi  KOMPAS.com - Jika Anda menggunaka...   
4     bisnis  Kompas  Tim Redaksi TANGERANG, KOMPAS.com - Siang hari...   
..       ...     ...                                                ...   
325  politik   Detik  KPK menilai jika mengimbau pelarangan pemberia...   
326  politik   Detik  Mahkamah Konstitusi (MK) menolak gugatan Anies...   
327  politik   Detik  Wali Kota Medan Bobby Nasution mengungkapkan k...   
328  politik   Detik  Ketua Harian Gerindra Dasco tak banyak membebe...   
329  politik   Detik  Ketua Umum PKB, Muhaimin Iskandar atau Cak Imi...   

                                          Cleaned Text  
0    pemerintah kabupaten sorong papua barat daya m...  
1    penulis jakarta kompascom  suara lakban yang d...  
2    sebuah galeri fesyen di tengah  aeon mall sent...  
3    jika anda menggunakan platform online untuk pe...  
4    siang hari di tengah panas yang terik orangora...  
..                                                 ...  
325  kpk menilai jika mengimbau pelarangan pemberia...  
326  mahkamah konstitusi mk menolak gugatan anies b...  
327  wali kota medan bobby nasution mengungkapkan k...  
328  ketua harian gerindra dasco tak banyak membebe...  
329  ketua umum pkb muhaimin iskandar atau cak imin...  

[330 rows x 4 columns]

Error: Runtime no longer has a reference to this dataframe, please re-run this cell and try again.
Error: Runtime no longer has a reference to this dataframe, please re-run this cell and try again.
Error: Runtime no longer has a reference to this dataframe, please re-run this cell and try again.
Error: Runtime no longer has a reference to this dataframe, please re-run this cell and try again.
Error: Runtime no longer has a reference to this dataframe, please re-run this cell and try again.


## Tokenization dan menghapus stop words

In [ ]:
# Inisialisasi stop words remover untuk bahasa Indonesia
stop_factory = StopWordRemoverFactory()
stopwords = stop_factory.get_stop_words()

def tokenize_text(text):
  tokens = word_tokenize(text)
  filtered_tokens = [word for word in tokens if word.lower() not in stopwords]
  return filtered_tokens

df_merged['Cleaned Text'] = df_merged['Cleaned Text'].apply(tokenize_text)
df_merged

Label   Media                                               Body  \
0     bisnis  Kompas  Editor  SORONG, KOMPAS.com - Pemerintah Kabupa...   
1     bisnis  Kompas  Penulis JAKARTA, KOMPAS.com – Suara lakban yan...   
2     bisnis  Kompas  Tim Redaksi  BOGOR, KOMPAS.com - Sebuah galeri...   
3     bisnis  Kompas  Tim Redaksi  KOMPAS.com - Jika Anda menggunaka...   
4     bisnis  Kompas  Tim Redaksi TANGERANG, KOMPAS.com - Siang hari...   
..       ...     ...                                                ...   
325  politik   Detik  KPK menilai jika mengimbau pelarangan pemberia...   
326  politik   Detik  Mahkamah Konstitusi (MK) menolak gugatan Anies...   
327  politik   Detik  Wali Kota Medan Bobby Nasution mengungkapkan k...   
328  politik   Detik  Ketua Harian Gerindra Dasco tak banyak membebe...   
329  politik   Detik  Ketua Umum PKB, Muhaimin Iskandar atau Cak Imi...   

                                          Cleaned Text  
0    [pemerintah, kabupaten, sorong, papua, barat, ...  
1    [penulis, jakarta, kompascom, suara, lakban, d...  
2    [sebuah, galeri, fesyen, tengah, aeon, mall, s...  
3    [menggunakan, platform, online, pemasaran, bis...  
4    [siang, hari, tengah, panas, terik, orangorang...  
..                                                 ...  
325  [kpk, menilai, mengimbau, pelarangan, pemberia...  
326  [mahkamah, konstitusi, mk, menolak, gugatan, a...  
327  [wali, kota, medan, bobby, nasution, mengungka...  
328  [ketua, harian, gerindra, dasco, tak, banyak, ...  
329  [ketua, umum, pkb, muhaimin, iskandar, cak, im...  

[330 rows x 4 columns]

## Stemming

In [ ]:
stemmer = StemmerFactory().create_stemmer()

def stem_tokens(tokens):
  stemmed_tokens = [stemmer.stem(token) for token in tokens]
  return stemmed_tokens

df_merged['Cleaned Text'] = df_merged['Cleaned Text'].apply(stem_tokens)
df_merged

Label   Media                                               Body  \
0     bisnis  Kompas  Editor  SORONG, KOMPAS.com - Pemerintah Kabupa...   
1     bisnis  Kompas  Penulis JAKARTA, KOMPAS.com – Suara lakban yan...   
2     bisnis  Kompas  Tim Redaksi  BOGOR, KOMPAS.com - Sebuah galeri...   
3     bisnis  Kompas  Tim Redaksi  KOMPAS.com - Jika Anda menggunaka...   
4     bisnis  Kompas  Tim Redaksi TANGERANG, KOMPAS.com - Siang hari...   
..       ...     ...                                                ...   
325  politik   Detik  KPK menilai jika mengimbau pelarangan pemberia...   
326  politik   Detik  Mahkamah Konstitusi (MK) menolak gugatan Anies...   
327  politik   Detik  Wali Kota Medan Bobby Nasution mengungkapkan k...   
328  politik   Detik  Ketua Harian Gerindra Dasco tak banyak membebe...   
329  politik   Detik  Ketua Umum PKB, Muhaimin Iskandar atau Cak Imi...   

                                          Cleaned Text  
0    [perintah, kabupaten, sorong, papua, barat, da...  
1    [tulis, jakarta, kompascom, suara, lakban, tar...  
2    [buah, galeri, fesyen, tengah, aeon, mall, sen...  
3    [guna, platform, online, pasar, bisnis, media,...  
4    [siang, hari, tengah, panas, terik, orangorang...  
..                                                 ...  
325  [kpk, nilai, imbau, larang, beri, dana, hibah,...  
326  [mahkamah, konstitusi, mk, tolak, gugat, anies...  
327  [wali, kota, medan, bobby, nasution, ungkap, i...  
328  [ketua, hari, gerindra, dasco, tak, banyak, be...  
329  [ketua, umum, pkb, muhaimin, iskandar, cak, im...  

[330 rows x 4 columns]

# Text Representation

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from gensim.models import Word2Vec
import numpy as np

## Metode 1: TF-IDF

TF-IDF, which stands for Term Frequency-Inverse Document Frequency, is a method for calculating the weight of a word in a document that reflects the importance of that word in that document relative to the data set.

In [ ]:
# Menggabungkan token yang telah distem ke dalam kalimat kembali
df_merged['Processed Text'] = df_merged['Cleaned Text'].apply(lambda x: ' '.join(x))

# Inisialisasi TF-IDF Vectorizer
tfidf_vectorizer = TfidfVectorizer()
tfidf_matrix = tfidf_vectorizer.fit_transform(df_merged['Processed Text'])

In [ ]:
# Mendapatkan nama fitur/kata dari TF-IDF Vectorizer
feature_names = tfidf_vectorizer.get_feature_names_out()

# Membuat DataFrame dari matriks TF-IDF
tfidf_df = pd.DataFrame(tfidf_matrix.toarray(), columns=feature_names)
tfidf_df

aamiin  aap  abadi  abai  abang  abdi  abdul  abdullah  abian  abis  ...  \
0       0.0  0.0    0.0   0.0    0.0   0.0    0.0       0.0    0.0   0.0  ...   
1       0.0  0.0    0.0   0.0    0.0   0.0    0.0       0.0    0.0   0.0  ...   
2       0.0  0.0    0.0   0.0    0.0   0.0    0.0       0.0    0.0   0.0  ...   
3       0.0  0.0    0.0   0.0    0.0   0.0    0.0       0.0    0.0   0.0  ...   
4       0.0  0.0    0.0   0.0    0.0   0.0    0.0       0.0    0.0   0.0  ...   
..      ...  ...    ...   ...    ...   ...    ...       ...    ...   ...  ...   
325     0.0  0.0    0.0   0.0    0.0   0.0    0.0       0.0    0.0   0.0  ...   
326     0.0  0.0    0.0   0.0    0.0   0.0    0.0       0.0    0.0   0.0  ...   
327     0.0  0.0    0.0   0.0    0.0   0.0    0.0       0.0    0.0   0.0  ...   
328     0.0  0.0    0.0   0.0    0.0   0.0    0.0       0.0    0.0   0.0  ...   
329     0.0  0.0    0.0   0.0    0.0   0.0    0.0       0.0    0.0   0.0  ...   

     zion  zirius  zita  zola  zon  zona  zoom  zuhur  zulhas  zulkifli  
0     0.0     0.0   0.0   0.0  0.0   0.0   0.0    0.0     0.0       0.0  
1     0.0     0.0   0.0   0.0  0.0   0.0   0.0    0.0     0.0       0.0  
2     0.0     0.0   0.0   0.0  0.0   0.0   0.0    0.0     0.0       0.0  
3     0.0     0.0   0.0   0.0  0.0   0.0   0.0    0.0     0.0       0.0  
4     0.0     0.0   0.0   0.0  0.0   0.0   0.0    0.0     0.0       0.0  
..    ...     ...   ...   ...  ...   ...   ...    ...     ...       ...  
325   0.0     0.0   0.0   0.0  0.0   0.0   0.0    0.0     0.0       0.0  
326   0.0     0.0   0.0   0.0  0.0   0.0   0.0    0.0     0.0       0.0  
327   0.0     0.0   0.0   0.0  0.0   0.0   0.0    0.0     0.0       0.0  
328   0.0     0.0   0.0   0.0  0.0   0.0   0.0    0.0     0.0       0.0  
329   0.0     0.0   0.0   0.0  0.0   0.0   0.0    0.0     0.0       0.0  

[330 rows x 7760 columns]

## Metode 2: Word2Vec

Word2Vec is a model that trains word vector representations from a collection of text. The model generates word vectors that reflect the semantic context of words; words that occur frequently in the same context have similar vector representations. Two architectures in word2vec:

- CBOW (Continuous Bag of Words)

  This model predicts the target word based on the surrounding words (context).

- Skip-Gram

  The opposite of CBOW, this model uses a single target word to predict surrounding words.

In [ ]:
sentences = df_merged['Cleaned Text'].tolist()

# Initialize and train the Word2Vec model
model_w2v = Word2Vec(sentences, vector_size=50, window=5, min_count=3, sg=1)

# Save and use the model
word_vectors = model_w2v.wv

i=0
for word in word_vectors.key_to_index:
  if(i>5): break
  vector = word_vectors[word]
  print(f"Word: {word} \nVector: {vector}\n")
  i+=1

Word: jadi 
Vector: [-0.12783508 -0.07230096  0.18290657 -0.12165687 -0.02577425  0.01635333
  0.08064151  0.14482574 -0.26775113 -0.15623222  0.18300042 -0.32221073
 -0.05682313  0.20848693  0.02541819  0.2165739   0.24547635  0.45662612
 -0.24369816 -0.28038573  0.7457555   0.35885838  0.2356802   0.42713654
  0.08450344  0.13370109 -0.20980349 -0.07453348 -0.37653616 -0.07953254
  0.21272346 -0.11282491  0.20835191  0.07944469 -0.20126465  0.23519918
  0.08442643 -0.3092436  -0.10210954  0.04168445 -0.06909011 -0.04996848
  0.06584734  0.14580145  0.36475933  0.29424948 -0.18188682 -0.32844785
  0.1367011   0.10166552]

Word: bisnis 
Vector: [-0.53183514 -0.28579864 -0.07634385 -0.21320492 -0.351092   -0.08734389
  0.1305756  -0.3864956  -0.17677294 -0.00480696  0.16856924 -0.53576434
  0.05347971 -0.06226281 -0.16777712  0.38481593  0.02020528  0.9910808
 -0.1613854  -0.57643884  0.6246186   1.1719027   0.19189224  0.44411358
  0.11303568  0.42418462  0.08236853 -0.10564256 -0.2044

In [ ]:
def document_vector(doc):
  # Menghapus kata yang tidak ada di model
  words = [word for word in doc if word in model_w2v.wv.key_to_index]
  if len(words) >= 1:
    return np.mean(model_w2v.wv[words], axis=0)
  else:
    return []

# Menghitung vektor rata-rata untuk setiap dokumen
df_merged['doc_vector'] = df_merged['Cleaned Text'].apply(document_vector)

# Menghapus baris yang tidak bisa di vektorisasi (jika ada)
df_merged = df_merged[df_merged['doc_vector'].map(len) > 0]
X_w2v = np.array(list(df_merged['doc_vector']))
X_w2v

array([[-1.14045593e-04, -2.53829043e-02,  1.16502926e-01, ...,
        -1.28784865e-01,  1.79427296e-01, -1.28635652e-02],
       [-4.00256366e-02, -6.41350001e-02,  1.80533737e-01, ...,
        -1.58075556e-01,  1.48990422e-01, -1.39467185e-02],
       [-6.26815483e-02, -8.61260295e-02,  2.39538372e-01, ...,
        -1.91906765e-01,  1.15424640e-01,  1.74701028e-02],
       ...,
       [-1.68566778e-02,  2.35411339e-02,  1.85247585e-01, ...,
        -2.22113222e-01,  9.02057216e-02,  8.57136026e-02],
       [-6.51945025e-02,  6.35096207e-02,  1.75786808e-01, ...,
        -2.28424892e-01,  1.30877839e-02,  2.40729332e-01],
       [ 1.17879948e-02,  1.17664307e-01,  1.35278150e-01, ...,
        -2.60591835e-01,  1.12343356e-01,  1.54271364e-01]], dtype=float32)

# Classification Model

In [ ]:
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.metrics import classification_report

## Train Test Split

In [ ]:
X_train_tfidf, X_test_tfidf, y_train, y_test = train_test_split(tfidf_matrix, df_merged['Label'], test_size=0.2, random_state=99)
X_train_w2v, X_test_w2v = train_test_split(X_w2v, test_size=0.2, random_state=99)

In [ ]:
len(X_w2v)

330

In [ ]:
# Mendefinisikan parameter untuk tuning
param_svm = {'C': [0.1, 1, 10], 'gamma': [0.001, 0.01, 0.1, 1]}
param_rf = {'n_estimators': [50, 100, 200], 'max_depth': [10, 20, 30]}

## SVM with TF-IDF

In [ ]:
svm_tfidf = GridSearchCV(SVC(), param_svm, cv=5)
svm_tfidf.fit(X_train_tfidf, y_train)
svm_tfidf_pred = svm_tfidf.predict(X_test_tfidf)

# Evaluasi model
svm_tfidf_report = classification_report(y_test, svm_tfidf_pred,  digits=5, output_dict=True)
print("SVM dengan TF-IDF:")
print(classification_report(y_test, svm_tfidf_pred,  digits=5))


SVM dengan TF-IDF:
              precision    recall  f1-score   support

      bisnis    0.80645   0.96154   0.87719        26
     hiburan    1.00000   0.55556   0.71429         9
    olahraga    0.90000   0.81818   0.85714        11
     politik    0.95000   0.95000   0.95000        20

    accuracy                        0.87879        66
   macro avg    0.91411   0.82132   0.84966        66
weighted avg    0.89194   0.87879   0.87370        66



## Random Forest with TF-IDF

In [ ]:
rf_tfidf = GridSearchCV(RandomForestClassifier(), param_rf, cv=5)
rf_tfidf.fit(X_train_tfidf, y_train)
rf_tfidf_pred = rf_tfidf.predict(X_test_tfidf)

rf_tfidf_report = classification_report(y_test, rf_tfidf_pred, digits = 5, output_dict=True)
print("Random Forest dengan TF-IDF:")
print(classification_report(y_test, rf_tfidf_pred, digits = 5))

Random Forest dengan TF-IDF:
              precision    recall  f1-score   support

      bisnis    0.92593   0.96154   0.94340        26
     hiburan    0.75000   0.66667   0.70588         9
    olahraga    0.90000   0.81818   0.85714        11
     politik    0.90476   0.95000   0.92683        20

    accuracy                        0.89394        66
   macro avg    0.87017   0.84910   0.85831        66
weighted avg    0.89120   0.89394   0.89161        66



## SVM with Word2Vec

In [ ]:
svm_w2v = GridSearchCV(SVC(), param_svm, cv=5)
svm_w2v.fit(X_train_w2v, y_train)
svm_w2v_pred = svm_w2v.predict(X_test_w2v)

svm_w2v_report = classification_report(y_test, svm_w2v_pred, digits=5, output_dict=True)
print("SVM dengan Word2Vec:")
print(classification_report(y_test, svm_w2v_pred, digits=5))

SVM dengan Word2Vec:
              precision    recall  f1-score   support

      bisnis    0.96154   0.96154   0.96154        26
     hiburan    1.00000   0.77778   0.87500         9
    olahraga    0.84615   1.00000   0.91667        11
     politik    0.95000   0.95000   0.95000        20

    accuracy                        0.93939        66
   macro avg    0.93942   0.92233   0.92580        66
weighted avg    0.94406   0.93939   0.93876        66



## Random Forest with Word2Vec

In [ ]:
rf_w2v = GridSearchCV(RandomForestClassifier(), param_rf, cv=5)
rf_w2v.fit(X_train_w2v, y_train)
rf_w2v_pred = rf_w2v.predict(X_test_w2v)

rf_w2v_report = classification_report(y_test, rf_w2v_pred, digits=5, output_dict=True)
print("Random Forest dengan Word2Vec:")
print(classification_report(y_test, rf_w2v_pred, digits=5))

Random Forest dengan Word2Vec:
              precision    recall  f1-score   support

      bisnis    0.96154   0.96154   0.96154        26
     hiburan    1.00000   0.77778   0.87500         9
    olahraga    0.84615   1.00000   0.91667        11
     politik    0.95000   0.95000   0.95000        20

    accuracy                        0.93939        66
   macro avg    0.93942   0.92233   0.92580        66
weighted avg    0.94406   0.93939   0.93876        66



# Performance Comparison

In [ ]:
# Create DataFrame
results = []

# Append results for TF-IDF SVM
results.append({
  'Text Representation': 'TF-IDF',
  'Algorithm': 'SVM',
  'Best Parameters': str(svm_tfidf.best_params_),
  'Accuracy': svm_tfidf_report['accuracy'],
  'Precision': svm_tfidf_report['macro avg']['precision'],
  'Recall': svm_tfidf_report['macro avg']['recall'],
  'F1 Score': svm_tfidf_report['macro avg']['f1-score']
})

# Append results for Random Forest with TF-IDF
results.append({
  'Text Representation': 'TF-IDF',
  'Algorithm': 'Random Forest',
  'Best Parameters': str(rf_tfidf.best_params_),
  'Accuracy': rf_tfidf_report['accuracy'],
  'Precision': rf_tfidf_report['macro avg']['precision'],
  'Recall': rf_tfidf_report['macro avg']['recall'],
  'F1 Score': rf_tfidf_report['macro avg']['f1-score']
})

# Append results for SVM with Word2Vec
results.append({
  'Text Representation': 'Word2Vec',
  'Algorithm': 'SVM',
  'Best Parameters': str(svm_w2v.best_params_),
  'Accuracy': svm_w2v_report['accuracy'],
  'Precision': svm_w2v_report['macro avg']['precision'],
  'Recall': svm_w2v_report['macro avg']['recall'],
  'F1 Score': svm_w2v_report['macro avg']['f1-score']
})

# Append results for Random Forest with Word2Vec
results.append({
  'Text Representation': 'Word2Vec',
  'Algorithm': 'Random Forest',
  'Best Parameters': str(rf_w2v.best_params_),
  'Accuracy': rf_w2v_report['accuracy'],
  'Precision': rf_w2v_report['macro avg']['precision'],
  'Recall': rf_w2v_report['macro avg']['recall'],
  'F1 Score': rf_w2v_report['macro avg']['f1-score']
})

# Create the final DataFrame
df_results = pd.DataFrame(results)
df_results

Text Representation      Algorithm                         Best Parameters  \
0              TF-IDF            SVM                 {'C': 10, 'gamma': 0.1}   
1              TF-IDF  Random Forest  {'max_depth': 10, 'n_estimators': 200}   
2            Word2Vec            SVM                    {'C': 1, 'gamma': 1}   
3            Word2Vec  Random Forest   {'max_depth': 30, 'n_estimators': 50}   

   Accuracy  Precision    Recall  F1 Score  
0  0.878788   0.914113  0.821319  0.849655  
1  0.893939   0.870172  0.849097  0.858313  
2  0.939394   0.939423  0.922329  0.925801  
3  0.939394   0.939423  0.922329  0.925801

- Word2Vec with SVM and Random Forest showed excellent and consistent performance across all metrics (accuracy, precision, recall, and F1 Score), with scores on all metrics around 0.92-0.94. This shows that the Word2Vec representation is very effective for this dataset, and both algorithms (SVM and Random Forest) are able to utilize it well.

- TF-IDF with SVM has a lower performance compared to using Word2Vec, with an accuracy of 0.878788, precision of 0.914113, recall of 0.821319, and F1 Score of 0.849655. Despite the high precision, the lower recall indicates that the model may not be as good at identifying all relevant positive cases.

- TF-IDF with Random Forest shows improvement compared to SVM in terms of recall and F1 Score, but slightly lower in precision. Accuracy was slightly higher at 0.893939, showing a general improvement compared to SVM.